### PR TITLE
Refactor marker file logic

### PR DESCRIPTION
**Pull request by Wyverald**
Marker files today store all predeclared inputs as one hash (on the first line of the file), and then each recorded input as a following line in the `TYPE:KEY VALUE` format. This commit refactors the parsing/stringification logic of recorded inputs so that they're not all clumped in big methods in `RepositoryFunction`, to pave the way for more recorded input types (watching directories, etc) and more places to write recorded input data (for the true repo cache).

- The StarlarkSemantics object is no longer treated as a recorded input (only recorded for Starlark repo rules, ignored for native repo rules), but as a predeclared input instead (i.e. hashed on the first line).
  - This slightly simplifies logic, and since the existing native repo rules are either local (local_repository, new_local_repository, local_config_platform) or being Starlarkified (the two Android repo rules), it will have minimal visible impact.
- Each type of recorded inputs is a subclass of `RepoRecordedInput`, which knows how to stringify itself, verify its own up-to-date-ness, and how to parse itself from a string.
- We also try to collect as many SkyKeys needed to verify up-to-date-ness as possible in one go and do a mass Skyframe evaluation. This avoids a fair amount of Skyframe restarts (unlikely to have super big impact on performance, but is a nice thing to do).

Work towards https://github.com/bazelbuild/bazel/issues/20952 and https://github.com/bazelbuild/bazel/issues/12227.

### CODE REVIEW

**Review by ahumesky — APPROVED** [review]

**Comment by fmeum** [comment]
I would go even further: I see value in aligning the repo marker file format with that of module extension lockfile entries. After all, there doesn't seem to be a conceptual reason for not supporting the same capabilities for module extensions eventually. We don't need to implement this now, but could switch over to Gson/JSON for serialization and possibly even use the same logic for verification of basic file accesses and recorded repo mapping entries.

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java [hunk 1]
```diff
@@ -643,22 +648,27 @@ private String readMarkerFile(String content, Map<String, String> markerData) {
...
       return markerRuleKey;
     }
 
-    private String computeRuleKey(Rule rule) {
+    private String computeRuleKey(Rule rule, StarlarkSemantics starlarkSemantics) {
       return new Fingerprint()
           .addBytes(RuleFormatter.serializeRule(rule).build().toByteArray())
           .addInt(MARKER_FILE_VERSION)
+          // TODO: Using the hashCode() method for StarlarkSemantics here is technically wrong.
+          //   hashCode() is not guaranteed to be stable across JVM restarts.
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java:670]: Only if using the default implementation of `hashCode()`, but `StarlarkSemantics` overrides it with logic that ultimately uses `hashCode()` on `ImmutableMap`, which is guaranteed to be stable (https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#hashCode--) as long as the `Object`s used as map values have stable hash code.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java:670]: But it's suboptimal for other reasons: It doesn't see changes to defaults. Could you add a comment about that?

**Comment by Wyverald** [comment]
it is indeed my plan to eventually switch to JSON. it's actually forwards compatible too, since the current marker file format treats the whole first line as the rule hash; so if the first line became `{`, older Bazel versions would simply think the rule hash changed and refetch.

And yeah, this PR paves the way to eventually merge this logic with the module extension up-to-date checks.

Otherwise, though, this PR is just a refactor -- it doesn't actually change the marker file format at all.

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 1 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: done. thanks I wasn't aware that Map had a documented hashcode algorithm (so does String, so it's actually completely stable).

**Review by meteorcloudy — APPROVED** [review]

**Review by fmeum — APPROVED** [review]

**Review by SalmaSamy — APPROVED** [review]

**Pull request closed without merging**